faer-libs/faer-core/src/complex_native/c32_impl.rs

```diff
@@ -0,0 +1,1051 @@
+        Self::new(self.re - rhs, self.im)
+    }
+}
+
+impl core::ops::Sub<c32> for f32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn sub(self, rhs: c32) -> Self::Output {
+        Self::Output::new(self - rhs.re, -rhs.im)
+    }
+}
+
+impl core::ops::Sub for c32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self::new(self.re - rhs.re, self.im - rhs.im)
+    }
+}
+
+impl core::ops::Sub<c32conj> for c32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn sub(self, rhs: c32conj) -> Self::Output {
+        Self::new(self.re - rhs.re, self.im + rhs.neg_im)
+    }
+}
+
+impl core::ops::Sub<c32> for c32conj {
+    type Output = c32;
+
+    #[inline(always)]
+    fn sub(self, rhs: c32) -> Self::Output {
+        Self::Output::new(self.re - rhs.re, -self.neg_im - rhs.im)
+    }
+}
+
+impl core::ops::Mul<f32> for c32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn mul(self, rhs: f32) -> Self::Output {
+        Self::new(self.re * rhs, self.im * rhs)
+    }
+}
+
+impl core::ops::Mul<c32> for f32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn mul(self, rhs: c32) -> Self::Output {
+        Self::Output::new(self * rhs.re, self * rhs.im)
+    }
+}
+
+impl core::ops::Mul for c32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn mul(self, rhs: Self) -> Self::Output {
+        Self::new(
+            self.re * rhs.re - self.im * rhs.im,
+            self.re * rhs.im + self.im * rhs.re,
+        )
+    }
+}
+
+impl core::ops::Mul<c32conj> for c32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn mul(self, rhs: c32conj) -> Self::Output {
+        Self::new(
+            self.re * rhs.re + self.im * rhs.neg_im,
+            self.im * rhs.re - self.re * rhs.neg_im,
+        )
+    }
+}
+
+impl core::ops::Mul<c32> for c32conj {
+    type Output = c32;
+
+    #[inline(always)]
+    fn mul(self, rhs: c32) -> Self::Output {
+        Self::Output::new(
+            self.re * rhs.re + self.neg_im * rhs.im,
+            rhs.im * self.re - rhs.re * self.neg_im,
+        )
+    }
+}
+
+impl core::ops::Div<f32> for c32 {
+    type Output = c32;
+
+    #[inline(always)]
+    fn div(self, rhs: f32) -> Self::Output {
+        Self::new(self.re / rhs, self.im / rhs)
+    }
+}
+
+impl core::ops::Div<c32> for f32 {
+    type Output = c32;
+
+    #[allow(clippy::suspicious_arithmetic_impl)]
+    #[inline(always)]
+    fn div(self, rhs: c32) -> Self::Output {
+        self * rhs.faer_inv()
+    }
+}
+
+impl core::ops::Div for c32 {
+    type Output = c32;
+
+    #[allow(clippy::suspicious_arithmetic_impl)]
+    #[inline(always)]
+    fn div(self, rhs: Self) -> Self::Output {
+        self * rhs.faer_inv()
+    }
+}
+
+impl core::ops::Div<c32conj> for c32 {
+    type Output = c32;
+
+    #[allow(clippy::suspicious_arithmetic_impl)]
+    #[inline(always)]
+    fn div(self, rhs: c32conj) -> Self::Output {
+        // TODO: Can be done a bit better
```

Comment:
i think the current implementation is fine, since division is a lot more expensive than canonicalization